op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+[monorepo PR 10051](https://github.com/ethereum-optimism/optimism/pull/10051);
+
+Pros:
+- The conditions/effect functions are easy to call manually, to perform sequences in non-flaky ways during testing.
+- It's easy to match the implementation with the specification rules.
+- Easily extensible/compositional. Adding `RepeatCond`, swapping conditions/effects, combining them, is easy.
+
+Cons:
+- The draft does not cleanly support prioritization of effects within a single resource.
+- It introduces more locks and `sync.Cond`, and doesn't follow the preference of communicating state.
+
+This is medium-effort, clears a decent amount of current tech-debt, but also introduces some new tech-debt.
+The pros are not exclusive to this solution, and the cons are.
+
+### Small state-loops
+
+This builds on the `RepeatCond` condition/effect idea, in particular the categorization by shared resource.
+There's an `unsafe`, `cross-unsafe`, `safe`, `finalized` and `currentL1` resource. 
+And each proceeds 1 at a time and can be signaled a reorg.
+
+When a condition is being checked, it implies a prestate for the effect, if the lock is not released in-between.
+By hosting the conditions/effects of the same resource together, it is easy to prioritize between them,
+and to keep a lock during execution of a condition/effect:
+the other conditions/effects would not be applicable while the primary resource itself is taken.
+
+The challenging thing here is to not copy-paste the issues of the current larger monolithic state-loop
+into these grouped smaller state-loops.
+
+Into this direction, we need to consider that:
+- When a resource is idle, it should be readable by many.
+- The loop itself should not inline key prioritization/scheduling logic, it makes it more difficult to test.
+
+The state-loops approach is not dynamically extensible: events are manually merged,
+prioritized, and categories of events are limited.
+
+This is medium-effort, but not all that much more testable than the current approach, and retains some tech-debt.
+
+### Derivers as block providers
+
+Another way to look at it, from @axelKingsley, is to think of it as a blocks-first system with "providers" and a "core".
+The "providers" produce blocks, and the "core" prioritizes and processes them, along with safety updates.
+
+There are some issues we identified:
+- Not all state changes are in fact blocks.
+- Creating blocks involves the engine (a contentious resource).
+- The "core" should not replicate the existing monolithic state-loop and synchronous execution issues.
+
+But this does start to introduce a plugin-like abstraction.
+A provider may do many things (alt-DA deriver) or just one thing (rollup safe-blocks deriver).
+Ideally we find a way such that the derivation, safety-progression and scheduling logic
+is modular such that we can just combine a few providers to build the driver.
+
+We believe we do not need runtime plugins as binaries:
```

Comment:
Ideally reth's builder api could be used as an inspiration for enabling a no-fork op-node with custom plugins